packages/fiber/src/core/stages.ts

```diff
@@ -0,0 +1,109 @@
+import { MutableRefObject } from 'react'
+import { StoreApi, UseBoundStore } from 'zustand'
+import { RootState } from './store'
+
+export interface UpdateCallback {
+  (state: RootState, delta: number, frame?: THREE.XRFrame | undefined): void
+}
+
+export type UpdateCallbackRef = MutableRefObject<UpdateCallback>
+type Store = UseBoundStore<RootState, StoreApi<RootState>>
+export type UpdateSubscription = { ref: UpdateCallbackRef; store: Store }
+
+export type FixedStageOptions = { fixedStep?: number; maxSubSteps?: number }
+
+export class Stage {
+  name: string
+  subscribers: UpdateSubscription[]
+
+  constructor(name: string) {
+    this.name = name
+    this.subscribers = []
+  }
+
+  frame(delta: number, frame?: THREE.XRFrame | undefined) {
+    const subs = this.subscribers
+
+    for (let i = 0; i < subs.length; i++) {
+      subs[i].ref.current(subs[i].store.getState(), delta, frame)
+    }
+  }
+
+  add(ref: UpdateCallbackRef, store: Store) {
+    this.subscribers = [...this.subscribers, { ref, store }]
+
+    return () => {
+      this.subscribers = this.subscribers.filter((sub) => {
+        return sub.ref !== ref
+      })
+    }
+  }
+}
+
+const FPS_50 = 1 / 50
+
+export class FixedStage extends Stage {
+  private fixedStep: number
+  private maxSubsteps: number
+  private accumulator: number
+  private alpha: number
+
+  constructor(name: string, options?: { fixedStep?: number; maxSubSteps?: number }) {
+    super(name)
+
+    this.fixedStep = options?.fixedStep ?? FPS_50
+    this.maxSubsteps = options?.maxSubSteps ?? 6
+    this.accumulator = 0
+    this.alpha = 0
+  }
+
+  frame(delta: number, frame?: THREE.XRFrame | undefined) {
+    const initialTime = performance.now()
+    let substeps = 0
+
+    this.accumulator += delta
+
+    while (this.accumulator >= this.fixedStep && substeps < this.maxSubsteps) {
+      this.accumulator -= this.fixedStep
+      substeps++
+
+      super.frame(this.fixedStep, frame)
+
+      if (performance.now() - initialTime > this.fixedStep * 200) {
```

Comment:
The static `200` here surprised me (since the total time checked will now entirely depend on the size of `fixedStep`.)

Reply:
Before the number was `1000`, which felt arbitrarily large so I made it `200` which I guess now feels arbitrarily small. My reasoning was if the loop is taking longer than 200 steps worth of time to complete, just stop. Maybe there is a better way that doesn't rely on arbitrary numbers. The purpose is to have an escape mechanism for possible spirals of death where the loop gets overwhelmed.